include/rmm/mr/device/detail/arena.hpp

```diff
@@ -0,0 +1,616 @@
+   * @return The (now freed) block associated with `p`. The caller is expected to return the block
+   * to the arena.
+   */
+  block free_block(void* p, std::size_t size) noexcept
+  {
+    auto const i = allocated_blocks_.find(p);
+    RMM_LOGGING_ASSERT(i != allocated_blocks_.end());
+
+    auto found = i->second;
+    RMM_LOGGING_ASSERT(found.size == size);
+    allocated_blocks_.erase(i);
+
+    return found;
+  }
+
+  /// The upstream resource to allocate memory from.
+  Upstream* upstream_mr_;
+  /// The maximum size the global arena can grow to.
+  std::size_t maximum_size_;
+  /// The current size of the global arena.
+  std::size_t current_size_{};
+  /// Address-ordered set of free blocks.
+  std::set<block> free_blocks_;
+  /// Map of pointer address to allocated blocks.
+  std::unordered_map<void*, block> allocated_blocks_;
+  /// Blocks allocated from upstream so that they can be quickly freed.
+  std::vector<block> upstream_blocks_;
+  /// Mutex for exclusive lock.
+  mutable std::mutex mtx_;
+};
+
+/**
+ * @brief An arena for allocating memory for a thread.
+ *
+ * An arena is a per-thread or per-non-default-stream memory pool for handling small-size
+ * allocations. It allocates superblocks from the global arena, and return them when the superblocks
+ * become empty.
+ *
+ * @tparam Upstream Memory resource to use for allocating the global arena. Implements
+ * rmm::mr::device_memory_resource interface.
+ */
+template <typename Upstream>
+class arena {
+ public:
+  /**
+   * @brief Construct an `arena`.
+   *
+   * @param global_arena The global arena from which to allocate superblocks.
+   */
+  explicit arena(global_arena<Upstream>& global_arena) : global_arena_{global_arena} {}
+
+  // Disable copy (and move) semantics.
+  arena(const arena&) = delete;
+  arena& operator=(const arena&) = delete;
+
+  /**
+   * @brief Allocates memory of size at least `bytes`.
+   *
+   * @throws `std::bad_alloc` if the requested allocation could not be fulfilled.
+   *
+   * @param bytes The size in bytes of the allocation.
+   * @return void* Pointer to the newly allocated memory.
+   */
+  void* allocate(std::size_t bytes)
+  {
+    lock_guard lock(mtx_);
+    auto const b = get_block(bytes);
+    allocated_blocks_.emplace(b.pointer, b);
+    return b.pointer;
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p`, and possibly return superblocks to upstream.
+   *
+   * @param p Pointer to be deallocated.
+   * @param bytes The size in bytes of the allocation. This must be equal to the value of `bytes`
+   * that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation.
+   * @return true if the allocation is found, false otherwise.
+   */
+  bool deallocate(void* p, std::size_t bytes, cudaStream_t stream)
+  {
+    lock_guard lock(mtx_);
+    bool found = do_deallocate(p, bytes);
+    shrink_arena(stream);
```

Comment:
shrink_arena is only ever called immediately after deleting a block. If that block is a superblock, it gets inserted into `free_superblocks` and then `shrink_arena` immedately frees it to upstream if there is more than one free superblock. So that means `free_superblocks` can only ever have either 0, 1, or 2 elements.  So, instead of keeping them in a `std::set`, why not just keep a single block called `free_superblock` and then, in `do_deallocate` just free any resulting superblock if there is already one in `free_superblock`, and if there isn't, just save it in `free_superblock`.  See related comments / suggestions.
```suggestion
```

Reply:
This is not completely true. When we are freeing memory allocated in another arena, we have to reach in, free the allocation, coalesce it, possibly generating superblocks, but since we don't have access to the stream associated with the other arena, we cannot safely return the superblocks. So it's possible for the size of `free_superblocks_` to be bigger.

Reply:
I missed that. This just points to the problem of inconsistently following stream-ordered-allocation semantics.